text/profiles/0237-profiles-data-model.md

```diff
@@ -0,0 +1,1015 @@
+message AttributeSet {
+  // Attributes associated with a specific Sample, Location or a Mapping.
+  // attributes is a collection of key/value pairs. Note, global attributes
+  // like server name can be set using the resource API. Examples of attributes:
+  //
+  //     "/http/user_agent": "Mozilla/5.0 (Macintosh; Intel Mac OS X 10_14_2) AppleWebKit/537.36 (KHTML, like Gecko) Chrome/71.0.3578.98 Safari/537.36"
+  //     "/http/server_latency": 300
+  //     "abc.com/myattribute": true
+  //     "abc.com/score": 10.239
+  //
+  // The OpenTelemetry API specification further restricts the allowed value types:
+  // https://github.com/open-telemetry/opentelemetry-specification/blob/main/specification/common/README.md#attribute
+  // Attribute keys MUST be unique (it is not allowed to have more than one
+  // attribute with the same key).
+  repeated opentelemetry.proto.common.v1.KeyValue attributes = 1;
+
+  // dropped_attributes_count is the number of attributes that were discarded. Attributes
+  // can be discarded because their keys are too long or because there are too many
+  // attributes. If this value is 0, then no attributes were dropped.
+  uint32 dropped_attributes_count = 2;
+}
+
+// A stacktrace is a sequence of locations. Order of locations goes from callers to callees. Many stacktraces will point to the same locations. The link between stacktraces, attribute sets, links, values and timestamps is implicit and is based on the order of the elements in the corresponding tables in ProfileType message.
+message Stacktrace {
+  repeated uint32 location_indices = 1;
+}
+
+// AggregationTemporality defines how a profile aggregator reports aggregated
+// values. It describes how those values relate to the time interval over
+// which they are aggregated.
+enum AggregationTemporality {
+  AGGREGATION_TEMPORALITY_UNSPECIFIED = 0; // Temporality unspecified.
+  AGGREGATION_TEMPORALITY_DELTA = 1;       // Delta aggregation over time.
+  AGGREGATION_TEMPORALITY_CUMULATIVE = 2;  // Cumulative aggregation over time.
+}
+
+// Represents a single profile type. It implicitly creates a connection between Stacktraces, Links, AttributeSets, values and timestamps. The connection is based on the order of the elements in the corresponding lists. This implicit connection creates an ephemeral structure called Sample. The length of reference lists must be the same. It is acceptable however for timestamps, links and attribute set lists to be empty. It is not acceptable for stacktrace or values lists to be empty.
+message ProfileType {
+  // aggregation_temporality describes if the aggregator reports delta changes
+  // since last report time, or cumulative changes since a fixed start time.
+  AggregationTemporality aggregation_temporality = 1;
+
+  // Profiler sample rate in Hz. This parameter indicates the frequency at which samples are collected, specifically for CPU profiles. Common values are 99 or 100. [Optional].
+  uint64 sample_rate = 2;
```

Comment:
Are we happy with 1Hz being the lowest frequency? Is there a use case when one would want to collect the samples at slower rate?

Reply:
>Common values are 99 or 100.

I am curious why this is so? Anything specially magical about these numbers or is it just customary in profiling world?

Reply:
> Are we happy with 1Hz being the lowest frequency? Is there a use case when one would want to collect the samples at slower rate?

A counter example would be a goroutine profile which is commonly taken once per profiling period (e.g. 60s). So I think we should change this. Maybe we should reconsider sticking with the [pprof approach](https://github.com/google/pprof/blob/main/proto/profile.proto#L82-L86C9) to this.

> > Common values are 99 or 100.
> 
> I am curious why this is so? Anything specially magical about these numbers or is it just customary in profiling world?

I think this value was popularized by Brendan Gregg, e.g. in his [acmque article](https://queue.acm.org/detail.cfm?id=2927301).

> stack traces are sampled at a rate of 99 times per second (not 100, to avoid lock-step sampling)